### PR TITLE
Fix Challenge Stats

### PR DESCRIPTION
**Pull request by Jayashsatolia403**
## Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
- blobber:
- gosdk:
- system_test:
- zboxcli:
- zwalletcli:
- Other: ...

### CODE REVIEW

File: code/go/0chain.net/core/maths/maths.go [hunk 1]
```diff
@@ -82,18 +82,18 @@ func SafeMultInt64(a, b int64) (int64, error) {
 }
 
 // ConvertToUint64 converts an int64 to uint64 and throws if the int64 number is negative
-func ConvertToUint64(a int64) (uint64, error) {
+func ConvertToUint64(a int64) (int64, error) {
```
Comment (dabasov) [inline code/go/0chain.net/core/maths/maths.go]: you can't change it, it is different function then

File: code/go/0chain.net/core/maths/maths.go [hunk 2]
```diff
@@ -82,18 +82,18 @@ func SafeMultInt64(a, b int64) (int64, error) {
...
-	return uint64(a), nil
+	return a, nil
 }
 
-// ConvertUInt64sToInts converts the given int64 numbers to corresponding unints and return them in the same order.
+// ConvertInt64sToUint64s converts the given int64 numbers to corresponding unints and return them in the same order.
 // throws if any value is less than 0.
-func ConvertUInt64sToInts(numbers ...int64) ([]uint64, error) {
-	convertedNumbers := make([]uint64, len(numbers))
+func ConvertInt64sToUint64s(numbers ...int64) ([]int64, error) {
```
Comment (dabasov) [inline code/go/0chain.net/core/maths/maths.go]: create different function

File: code/go/0chain.net/smartcontract/dbs/event/blobber.go [hunk 3]
```diff
@@ -411,46 +411,61 @@ func withBlobberStatsMerged() eventMergeMiddleware {
...
+		a.CompletedDelta += b.CompletedDelta
+		a.PassedDelta += b.PassedDelta
+		a.OpenDelta += b.OpenDelta
+		return a, nil
+	})
+}
 
-	return CreateBuilder("blobbers", "id", blobberIdList).
-		AddUpdate("open_challenges", openChallengesList).Exec(edb).Error
+func mergeAddChallengesToBlobberEvents() *eventsMergerImpl[ChallengeStatsDeltas] {
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/dbs/event/blobber.go]: why did you remove the builder?

File: code/go/0chain.net/smartcontract/dbs/event/blobber.go [hunk 4]
```diff
@@ -411,46 +411,61 @@ func withBlobberStatsMerged() eventMergeMiddleware {
...
+func (edb *EventDb) updateOpenBlobberChallenges(deltas []ChallengeStatsDeltas) error {
+	return edb.Store.Get().Raw(sqlUpdateOpenChallenges(deltas)).Scan(&Blobber{}).Error
+}
 
-	for _, blobber := range blobbers {
-		blobberIdList = append(blobberIdList, blobber.ID)
-		challengesPassedList = append(challengesPassedList, blobber.ChallengesPassed)
-		challengesCompletedList = append(challengesCompletedList, blobber.ChallengesCompleted)
-		challengesOpenList = append(challengesOpenList, blobber.OpenChallenges)
+func sqlUpdateOpenChallenges(deltas []ChallengeStatsDeltas) string {
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/dbs/event/blobber.go]: use builder here, do not use raw queries

File: code/go/0chain.net/smartcontract/dbs/event/blobber.go [hunk 5]
```diff
@@ -505,6 +520,34 @@ func (edb *EventDb) blobberSpecificRevenue(spus []dbs.StakePoolReward) error {
 		Exec(edb).Debug().Error
 }
 
+// ref https://www.postgresql.org/docs/9.1/sql-values.html
+func sqlUpdateBlobberChallenges(deltas []ChallengeStatsDeltas) string {
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/dbs/event/blobber.go]: use builder here, you can't use custom queries

**Review by dabasov — CHANGES_REQUESTED** [review]

**Comment by Jayashsatolia403** [comment]
Closed accidently. Really sorry for inconvenience.

**Review by dabasov — APPROVED** [review]
lgtm

**Pull request merged**